### PR TITLE
[radio] add supported and preferred channel mask

### PR DESCRIPTION
**Pull request by chshu**
Add new radio platform APIs:
 * otPlatRadioGetSupportedChannelMask(); // supported channel mask: the device is allowed to be on;
 * otPlatRadioGetPreferredChannelMask(); // preferred channel mask: the device prefers to form on.

Remove unused otLinkGetPhy*() APIs.

Add spinel property `SPINEL_PROP_PHY_CHAN_PREFERRED` to retrieve Preferred ChannelMask.

### CODE REVIEW

File: include/openthread/platform/radio.h [hunk 1]
```diff
@@ -694,6 +694,22 @@ void otPlatRadioClearSrcMatchShortEntries(otInstance *aInstance);
  */
 void otPlatRadioClearSrcMatchExtEntries(otInstance *aInstance);
 
+/**
+ * Get the radio supported channel mask that the device is allowed to be on.
+ *
+ * @param[in]  aInstance   The OpenThread instance structure.
+ *
+ */
+uint32_t otPlatRadioGetSupportedChannelMask(otInstance *aInstance);
```
Comment (abtink) [inline include/openthread/platform/radio.h]: How about instead of `OPENTHREAD_CONFIG_RADIO_CHANNEL_MASK` we use a radio capability from `otRadioCaps` definition to indicate whether the radio platform supports these APIs or not?

File: src/ncp/spinel.c [hunk 2]
```diff
@@ -1271,6 +1271,10 @@ const char *spinel_prop_key_to_cstr(spinel_prop_key_t prop_key)
         ret = "PHY_CHAN_SUPPORTED";
         break;
 
+    case SPINEL_PROP_PHY_CHAN_PREFERRED:
```
Comment (abtink) [inline src/ncp/spinel.c]: nit: We usually keep the order of `case` statements here match the order in `spinel.h` definition (make it easier to check if one is missing later). How about moving this after `SPINEL_PROP_PHY_PCAP_ENABLED`?

File: src/core/mac/mac.hpp [hunk 3]
```diff
@@ -341,6 +341,14 @@ class Mac : public InstanceLocator, public SubMac::Callbacks
      */
     void SetSupportedChannelMask(const ChannelMask &aMask);
 
+    /**
+     * This method returns the phy supported channel mask.
+     *
+     * @returns The phy supported channel mask.
+     *
+     */
+    uint32_t GetPhySupportedChannelMask(void) const;
```
Comment (abtink) [inline src/core/mac/mac.hpp]: Suggestion: How about making this new method a private one and allowing others to use `GetSupportedChannelMask()`. This way, the supported channel mask would be configurable by both radio platform and user (through the public API `otLinkSetSupportedChannelMask`).

File: src/core/meshcop/dataset_manager.cpp [hunk 4]
```diff
@@ -227,7 +227,7 @@ otError DatasetManager::GetChannelMask(Mac::ChannelMask &aChannelMask) const
     VerifyOrExit(channelMaskTlv != NULL, error = OT_ERROR_NOT_FOUND);
     VerifyOrExit((mask = channelMaskTlv->GetChannelMask()) != 0);
 
-    aChannelMask.SetMask(mask & Phy::kSupportedChannels);
+    aChannelMask.SetMask(mask & Get<Mac::Mac>().GetPhySupportedChannelMask());
```
Comment (abtink) [inline src/core/meshcop/dataset_manager.cpp]: How about using `GetSupportedChannelMask()` here instead?

File: include/openthread/platform/radio.h (hunk 1 above)
Comment (chshu) [reply include/openthread/platform/radio.h]: otRadioCaps is a run time capacity check, if we use otRadioCaps instead, every platform needs to implement the APIs (real implementation or at least a stub), it make the APIs like a mandatory  feature, but not optional as designed. Do we want to make it a mandatory feature?

File: src/core/mac/mac.hpp (hunk 3 above)
Comment (chshu) [reply src/core/mac/mac.hpp]: We have another otLinkGetPhySupportedChannelMask() will call it, so keep it as public. Generally, the SupportedChannelMask and PhySupportedChannelMask could be different. User can't change the PhySupportedChannelMask, but can change the SupportedChannelMask.

File: src/core/mac/mac.cpp [hunk 5]
```diff
@@ -435,6 +435,15 @@ void Mac::SetSupportedChannelMask(const ChannelMask &aMask)
     return;
 }
 
+uint32_t Mac::GetPhySupportedChannelMask(void) const
+{
+#if OPENTHREAD_CONFIG_RADIO_CHANNEL_MASK
```
Comment (bukepo) [inline src/core/mac/mac.cpp]: Is it possible to provide a weak implementation for so `otPlatRadioGetSupportedChannelMask()` that we don't need this macro.

File: src/ncp/ncp_base.cpp [hunk 6]
```diff
@@ -1793,6 +1793,10 @@ template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_CAPS>(void)
     SuccessOrExit(error = mEncoder.WriteUintPacked(SPINEL_CAP_SLAAC));
 #endif
 
+#if OPENTHREAD_CONFIG_RADIO_CHANNEL_MASK
+    SuccessOrExit(error = mEncoder.WriteUintPacked(SPINEL_CAP_RADIO_CHANNEL_MASK));
```
Comment (bukepo) [inline src/ncp/ncp_base.cpp]: Is it possible to avoid introducing this new capability? If this feature is not supported, we have simply return not found error when getting or setting `SPINEL_PROP_PHY_CHAN_PREFERRED`. I'm not in favor of resulting in hundreds of caps.

Comment (chshu) [reply src/ncp/ncp_base.cpp]: Agree that we can get rid of this new cap if we can accept the not found error, wpantund can set the preferred channel mask to 0 when not found.

@abtink, May I know your opinion?

File: src/core/mac/mac.cpp (hunk 5 above)
Comment (chshu) [reply src/core/mac/mac.cpp]: This macro will also help to reduce code size when Radio channel mask related features are not enabled.

File: include/openthread/platform/radio.h (hunk 1 above)
Comment (abtink) [reply include/openthread/platform/radio.h]: Good point on radio capability. 

I was also thinking of adding weak implementation (noticed @bukepo suggested the same idea). 
- We can add `OT_TOOL_WEAK` default implementation for both `otPlatRadioGetSupportedChannekMask` and `otPlatRadioGetPreferredChannelMask()` and radio platform code can choose to override either definition and provide its own implementation. 
  - We can make default implementation of `otPlatRadioGetSupportedChannekMask()` to return `Phy::kSupportedChannels`
  - And default implementation of `otPlatRadioGetPreferredChannelMask()` actually returning the `otPlatRadioGetSupportedChannelMask()` (this way radio platform can choose to just override the `GetSupportedChannelMask()`)

- I feel a weak implementation of the API may be simpler and cleaner (no need for a new feature `OPENTHREAD_CONFIG_RADIO_CHANNEL_MASK` and/or the spinel capability). 
- I understand your point  that having `OPENTHREAD_CONFIG_RADIO_CHANNEL_MASK` would reduce code size but this feature is not adding a lot of code and is not that big, so may be fine to have it be available on all builds. I perosnally think simplicity of design may be more preferable in this case over the small code size saving.

Thoughts?

File: src/core/mac/mac.hpp (hunk 3 above)
Comment (abtink) [reply src/core/mac/mac.hpp]: I am thinking that we don't expose `otLinkGetPhySupportedChannelMask()` as a public API.

Basically, from OT perspective, we allow supported channel mask to be specified from higher layer (through public API `otLinkSetSupportedChannelMask`) and/or from radio layer (through the new `otPlatRadioGetSupportedChannelMask()` API).

In `Mac`  we keep track of the mask which is the combination/intersection of the two (from public API and radio API) and everywhere else in OT core we use the `Mac::GetSupportedChannelMask()`.  

The public API `otLinkGetSupportedChannelMask()` would also get the mask from `Mac` (which it already does) so it would be always the mask that is being used by OT core.

Thoughts?

File: src/core/thread/announce_sender.cpp [hunk 7]
```diff
@@ -69,7 +69,7 @@ otError AnnounceSenderBase::SendAnnounce(Mac::ChannelMask aChannelMask,
     VerifyOrExit(aPeriod != 0, error = OT_ERROR_INVALID_ARGS);
     VerifyOrExit(aJitter < aPeriod, error = OT_ERROR_INVALID_ARGS);
 
-    aChannelMask.Intersect(Mac::ChannelMask(Phy::kSupportedChannels));
+    aChannelMask.Intersect(Mac::ChannelMask(Get<Mac::Mac>().GetSupportedChannelMask().GetMask()));
```
Comment (abtink) [inline src/core/thread/announce_sender.cpp]: I think we may be able to simplify the code like this: 
`aChannelMask.Intersect(Get<Mac::Mac>().GetSupportedChannelMask()))`

File: src/ncp/ncp_base_mtd.cpp [hunk 8]
```diff
@@ -207,11 +207,6 @@ template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_PHY_FREQ>(void)
     return mEncoder.WriteUint32(freq_khz);
 }
 
-template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_PHY_CHAN_SUPPORTED>(void)
-{
-    return EncodeChannelMask(otLinkGetPhySupportedChannelMask(mInstance));
```
Comment (abtink) [inline src/ncp/ncp_base_mtd.cpp]: This does not seem correct to me? (not what I remember/expect from the earlier code).
The previous implementation should have been using the `otLinkGetSupportedChannelMask()`? (which maps to `Mac::GetSupportedChannelMask()`. The same way the setter is using `otLinkSetSupportedChannelMask`.

Comment (abtink) [reply src/ncp/ncp_base_mtd.cpp]: Looks like it was changed by this commit (probably a mistake). 
https://github.com/openthread/openthread/commit/96725a985ed57a64721c61c4721e5792b44264c5#diff-523683496c0ec331f8caa4137d95295b

File: src/ncp/ncp_base.cpp [hunk 9]
```diff
@@ -2173,6 +2177,24 @@ template <> otError NcpBase::HandlePropertySet<SPINEL_PROP_DEBUG_NCP_LOG_LEVEL>(
     return error;
 }
 
+#if !OPENTHREAD_RADIO || OPENTHREAD_CONFIG_RADIO_CHANNEL_MASK
+template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_PHY_CHAN_SUPPORTED>(void)
```
Comment (abtink) [inline src/ncp/ncp_base.cpp]: How about this for getter for `SPINEL_PROP_PHY_CHAN_SUPPORTED`?
- On FTD and MTD, we encode `otLinkGetSupportedChannelMask()`
- On RADIO, we encode `otPlatRadioGetSupportedChannelMask()`

File: include/openthread/platform/radio.h (hunk 1 above)
Comment (chshu) [reply include/openthread/platform/radio.h]: Thanks for your suggestions, it's ok for me to remove the macro definition with little code size increasement, and agree that weak method is much simpler. The PR is updated accordingly.

File: src/core/mac/mac.hpp (hunk 3 above)
Comment (chshu) [reply src/core/mac/mac.hpp]: Agree, I moved the `otLinkGetPhySupportedChannelMask()` (as well as two other otLinkGetPhy*()) to be radio platform API, and have a weak implementation for them. Then in this case, `Mac::GetPhySupportedChannelMask()` is not necessary more, removed it.

File: include/openthread/platform/radio.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (chshu) [reply include/openthread/platform/radio.h]: Yes, just updated to include the change.

[the remark this answers is not in the document]
Comment (zhanglongxia) [reply include/openthread/platform/radio.h]: The function `otLinkGetPhyChannelMin` and `otLinkGetPhyChannelMax` may also should be provided by radio driver.

File: src/ncp/ncp_base_mtd.cpp (hunk 8 above)
Comment (chshu) [reply src/ncp/ncp_base_mtd.cpp]: Yes, I have fixed it within this PR.

File: include/openthread/platform/radio.h [hunk 10]
```diff
@@ -694,6 +694,46 @@ void otPlatRadioClearSrcMatchShortEntries(otInstance *aInstance);
...
+
+/**
+ * Get the minimum channel number of IEEE 802.15.4 physical layer.
+ *
+ * @param[in]  aInstance   The OpenThread instance strucyyture.
+ *
+ * @returns The minimum channel number.
+ *
+ */
+uint8_t otPlatRadioGetPhyChannelMin(otInstance *aInstance);
```
Comment (bukepo) [inline include/openthread/platform/radio.h]: I think we should submit a separate PR to remove these two functions. Just use the channel mask for all cases. I think we should reduce APIs.

File: src/cli/cli.cpp [hunk 11]
```diff
@@ -560,8 +560,8 @@ void Interpreter::ProcessChannel(int argc, char *argv[])
                 mServer->OutputFormat("count: %lu\r\n", otChannelMonitorGetSampleCount(mInstance));
 
                 mServer->OutputFormat("occupancies:\r\n");
-                for (uint8_t channel = otLinkGetPhyChannelMin(mInstance); channel <= otLinkGetPhyChannelMax(mInstance);
-                     channel++)
+                for (uint8_t channel = otPlatRadioGetPhyChannelMin(mInstance);
```
Comment (bukepo) [inline src/cli/cli.cpp]: again, we should update this to use channel mask directly.

File: src/posix/platform/radio_spinel.cpp [hunk 12]
```diff
@@ -1337,6 +1337,37 @@ otError RadioSpinel::PlatDiagProcess(const char *aString, char *aOutput, size_t
...
+    SuccessOrExit(error = Get(aKey, SPINEL_DATATYPE_DATA_S, maskBuffer, &maskLength));
+
+    while (maskLength > 0)
+    {
+        uint8_t        channel;
+        spinel_ssize_t unpacked;
+
+        unpacked = spinel_datatype_unpack(maskData, maskLength, SPINEL_DATATYPE_UINT8_S, &channel);
+        VerifyOrExit(unpacked > 0, error = OT_ERROR_FAILED);
+        VerifyOrExit(channel < kChannelMaskBufferSize, error = OT_ERROR_INVALID_ARGS);
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: #define as `kMaxChannel`? And maybe we should just drop unsupported channel or return parse error.

File: src/posix/platform/radio_spinel.hpp [hunk 13]
```diff
@@ -439,6 +439,18 @@ class RadioSpinel : public HdlcInterface::Callbacks
     otError PlatDiagProcess(const char *aString, char *aOutput, size_t aOutputMaxLen);
 #endif
 
+    /**
+     * This method returns the radio channel mask.
+     *
+     * @param[in]   aKey  Spinel property key (SPINEL_PROP_PHY_CHAN_SUPPORTED or SPINEL_PROP_PHY_CHAN_PREFERRED).
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.hpp]: prefer
```
* @param[in]   aPreferred  TRUE to get preferred channel mask, FALSE to get supported channel mask.
```
or define a specific enum for this purpose.

File: include/openthread/platform/radio.h (hunk 10 above)
Comment (chshu) [reply include/openthread/platform/radio.h]: Agree, deleted the two APIs.

File: src/posix/platform/radio_spinel.cpp (hunk 12 above)
Comment (chshu) [reply src/posix/platform/radio_spinel.cpp]: Here the kChannelMaskBufferSize (32) is different from kMaxChannel (normally 26). Updated to OT_ERROR_PARSE.

File: src/cli/cli.cpp [hunk 14]
```diff
@@ -560,8 +560,7 @@ void Interpreter::ProcessChannel(int argc, char *argv[])
                 mServer->OutputFormat("count: %lu\r\n", otChannelMonitorGetSampleCount(mInstance));
 
                 mServer->OutputFormat("occupancies:\r\n");
-                for (uint8_t channel = otLinkGetPhyChannelMin(mInstance); channel <= otLinkGetPhyChannelMax(mInstance);
-                     channel++)
+                for (uint8_t channel = Phy::kChannelMin; channel <= Phy::kChannelMax; channel++)
```
Comment (jwhui) [inline src/cli/cli.cpp]: `Phy::kChannelMin` and `Phy::kChannelMax` are not defined in public header files.

File: src/ncp/ncp_base_mtd.cpp [hunk 15]
```diff
@@ -2067,7 +2062,7 @@ template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_CHANNEL_MONITOR_CHANN
 {
     otError error = OT_ERROR_NONE;
 
-    for (uint8_t channel = otLinkGetPhyChannelMin(mInstance); channel <= otLinkGetPhyChannelMax(mInstance); channel++)
+    for (uint8_t channel = Phy::kChannelMin; channel <= Phy::kChannelMax; channel++)
```
Comment (jwhui) [inline src/ncp/ncp_base_mtd.cpp]: Should avoid directly referencing internal symbols when possible.

File: src/posix/platform/radio_spinel.hpp [hunk 16]
```diff
@@ -439,6 +439,18 @@ class RadioSpinel : public HdlcInterface::Callbacks
...
+     * This method returns the radio channel mask.
+     *
+     * @param[in]   aPreferred  TRUE to get preferred channel mask, FALSE to get supported channel mask.
+     *
+     * @returns The radio channel mask according to @aPreferred:
+     *   The radio supported channel mask that the device is allowed to be on.
+     *   The radio preferred channel mask that the device prefers to form on.
+     *
+     */
+    uint32_t GetRadioChannelMask(bool aPreferred);
```
Comment (jwhui) [inline src/posix/platform/radio_spinel.hpp]: Declare as `const`?

File: src/cli/cli.cpp (hunk 14 above)
Comment (chshu) [reply src/cli/cli.cpp]: Ok, I see. So `otLinkGetPhyChannelMin()` and `otLinkGetPhyChannelMax()` APIs are still required here. Reverted.

File: src/ncp/ncp_base_mtd.cpp (hunk 15 above)
Comment (chshu) [reply src/ncp/ncp_base_mtd.cpp]: Reverted, still use `otLinkGetPhyChannelMin()` and `otLinkGetPhyChannelMax()`.

File: src/posix/platform/radio_spinel.hpp (hunk 16 above)
Comment (chshu) [reply src/posix/platform/radio_spinel.hpp]: We can't declare as const here, since this function will call into RadioSpinel::Get() which will change the member variables.

File: src/cli/cli.cpp (hunk 14 above)
Comment (jwhui) [reply src/cli/cli.cpp]: Is it possible to use the channel mask directly?

Comment (chshu) [reply src/cli/cli.cpp]: I saw some behavior changes between using ChannelMask VS ChannelMin/Max. If we use channelmask, if will start from 0 every time, and only handle the supported channels.

If it is acceptable, can I submit a separate PR to remove the `otLinkGetPhyChannelMin()` and `otLinkGetPhyChannelMax()` related usage? So to keep this PR only focus on adding preferred channelmask feature.
